### PR TITLE
[ALLUXIO-2039] Fixing mounting of Swift directories.

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses: https://alluxio.atlassian.net/browse/ALLUXIO-2039

### CODE REVIEW

**Comment by madanadit** [comment]
LGTM

**Comment by ifcharming** [comment]
LGTM

As a separate note, maybe we should come up with better integration tests for various UFSs.

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 1]
```diff
@@ -264,14 +270,13 @@ public long getSpace(String path, SpaceType type) throws IOException {
 
   @Override
   public boolean isFile(String path) throws IOException {
-    String strippedPath = stripPrefixIfPresent(path);
-    return mAccount.getContainer(mContainerName).getObject(strippedPath).exists();
+    return exists(path) && !isDirectory(path);
```
Comment (gpang) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: `isFile` calls `exists` then `isDirectory`. They both will call swift apis, which means it now does multiple rpcs. The previous code looks like it does only 1? Is this expected?

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 2]
```diff
@@ -387,39 +392,37 @@ private boolean copy(String src, String dst) {
...
    * @param path the path to check
    * @return boolean indicating if the path is a directory
    * @throws IOException if an error occurs listing the directory
    */
   private boolean isDirectory(String path) throws IOException {
-    String strippedPath = stripPrefixIfPresent(path);
-    String[] children = listInternal(strippedPath, true);
-    return children != null && children.length > 0;
+    String[] children = listInternal(path);
+    return children.length > 0;
```
Comment (gpang) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: Why remove the `null` check?

**Comment by gpang** [comment]
@jsimsa Thanks for the fix! I left a few questions.

Also, what is the root cause for the mounting issue?

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 3]
```diff
@@ -208,7 +208,13 @@ public boolean delete(String path, boolean recursive) throws IOException {
   public boolean exists(String path) throws IOException {
     // To get better performance Swift driver does not create a _temporary folder.
     // This optimization should be hidden from Spark, therefore exists _temporary will return true.
-    return path.endsWith("_temporary") || isFile(path) || isDirectory(path);
+    if (path.endsWith("_temporary")) {
+      return true;
+    }
+    String strippedPath = stripPrefixIfPresent(path);
+    Container c = mAccount.getContainer(mContainerName);
```
Comment (apc999) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: NIT: `c` -> `container` ?
and `so` -> `object` or `storedObject`,
to be more consistent with our var naming style

**Comment by apc999** [comment]
@jsimsa I left a minor comment, I don't feel strong about it, I will leave it up to you. Otherwise, LGTM

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java (hunk 3 above)
Comment (jsimsa) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: changed everywhere in this file

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java (hunk 2 above)
Comment (jsimsa) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: fixed

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java (hunk 1 above)
Comment (jsimsa) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: FWIW, this is consistent with what we do for other object storage (GCS, S3, OSS). I think the motivation behind this is to optimize for the case where the path does not exist. However, as you correctly point out that it makes the case where the path exists more expensive. I think the right thing to do is to actually avoid reusing exists and isDirectory and instead write a custom function, which is what I will do.

Comment (jsimsa) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: as per our offline discussion, I will leave it as is for the time being and address this across all object storage UFSes in a separate PR

**Comment by jsimsa** [comment]
@apc999 @gpang addressed your comments, PTAL

**Comment by gpang** [comment]
LGTM

**Comment by apc999** [comment]
LGTM

**Pull request merged**